### PR TITLE
[Serve] Recover PENDING_INITIALIZATION status actor

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

Then replica is under initializing state and the controller is dead, user will see 
```
(ServeController pid=458318) ERROR 2023-03-29 08:44:42,547 controller 458318 deployment_state.py:500 - Exception in deployment 'xqHWInctmH'
(ServeController pid=458318) Traceback (most recent call last):
(ServeController pid=458318)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/serve/_private/deployment_state.py", line 489, in check_ready
(ServeController pid=458318)     deployment_config, version = ray.get(self._ready_obj_ref)
(ServeController pid=458318)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
(ServeController pid=458318)     return func(*args, **kwargs)
(ServeController pid=458318)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/_private/worker.py", line 2508, in get
(ServeController pid=458318)     raise value.as_instanceof_cause()
(ServeController pid=458318) ray.exceptions.RayTaskError(AttributeError): ray::ServeReplica:xqHWInctmH.get_metadata() (pid=458148, ip=172.31.126.222, repr=<ray.serve._private.replica.ServeReplica:xqHWInctmH object at 0x7f57ae0f0110>)
(ServeController pid=458318)   File "/home/ray/anaconda3/lib/python3.7/concurrent/futures/_base.py", line 428, in result
(ServeController pid=458318)     return self.__get_result()
(ServeController pid=458318)   File "/home/ray/anaconda3/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
(ServeController pid=458318)     raise self._exception
(ServeController pid=458318)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/serve/_private/replica.py", line 249, in get_metadata
(ServeController pid=458318)     return self.replica.deployment_config, self.replica.version
(ServeController pid=458318) AttributeError: 'NoneType' object has no attribute 'deployment_config'
```

- Fix the NoneType bug when recover happens, no matter the actor is under any state, user would not see the traceback. Instead user will see the 
1.  slow startup warning, and then replica will be terminated, and new replica will be provisioned.
2. If actor succeed in `PENDING_INITIALIZATION`, no error pops out.

This is observed from long_running_serve_failure: https://console.anyscale-staging.com/o/anyscale-internal/projects/prj_qC3ZfndQWYYjx2cz8KWGNUL4/clusters/ses_u7xeve33e2djg9grgr9qcs9l4x?command-history-section=command_history

**Note**: If user constructor initialization just needs very long time to finish, it is recommended to increase the `SERVE_SLOW_STARTUP_WARNING_S`, (if the user doesn't change, the deployment manager will terminate the old replica and start new replica after the timeout.) 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/test_controller_recovery.py [hunk 1]
```diff
@@ -200,5 +200,48 @@ def make_nonblocking_calls(expected, expect_blocking=False, num_returns=1):
...
+    @ray.remote
+    def pending_init_indicator():
+        ray.get(signal2.wait.remote())
+        return True
+
+    @serve.deployment
+    class V1:
+        def __init__(self):
+            signal2.send.remote()
+            ray.get(signal.wait.remote())
```
Comment (zcin) [inline python/ray/serve/tests/test_controller_recovery.py:219]:
```suggestion
            await signal.wait.remote()
```

File: python/ray/serve/_private/deployment_state.py [hunk 2]
```diff
@@ -497,7 +497,10 @@ def check_ready(self) -> Tuple[ReplicaStartupStatus, Optional[DeploymentVersion]
                     self._allocated_obj_ref
                 )
             except Exception:
-                logger.exception(f"Exception in deployment '{self._deployment_name}'")
+                logger.exception(
+                    f"Exception in replica '{self._replica_tag}',"
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:501]:
```suggestion
                    f"Exception in replica '{self._replica_tag}', "
```

File: python/ray/serve/_private/replica.py [hunk 3]
```diff
@@ -243,9 +243,16 @@ async def reconfigure(
...
-            return self.get_metadata()
+            return await self.get_metadata()
 
-        def get_metadata(self) -> Tuple[DeploymentConfig, DeploymentVersion]:
+        async def get_metadata(
+            self,
+        ) -> Tuple[DeploymentConfig, DeploymentVersion]:
+            # When the self.replica variable is None, potentially replica is still
+            # under initialization. Relying on DeploymentStateManager to
+            # timeout the replica.
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:253]: I'm a little lost at why this change is needed. Is it because the constructor might still be running while we are recovering?

Comment (edoakes) [reply python/ray/serve/_private/replica.py:253]: If so, can we use an explicit `asyncio.Event` instead of implementing our own sleeping logic here?

Comment (edoakes) [reply python/ray/serve/_private/replica.py:253]: e.g., `self._initialized = asyncio.Event()`

Comment (sihanwang41) [reply python/ray/serve/_private/replica.py:253]: > I'm a little lost at why this change is needed. Is it because the constructor might still be running while we are recovering?
yes

super, I forget this Async variable! updated.

File: python/ray/serve/_private/replica.py [hunk 4]
```diff
@@ -243,9 +243,16 @@ async def reconfigure(
...
+            while self.replica is None:
+                await asyncio.sleep(0.2)
```
[the remark this answers is not in the document]
Comment (sihanwang41) [reply python/ray/serve/_private/replica.py:255]: updated.

File: python/ray/serve/tests/test_controller_recovery.py [hunk 5]
```diff
@@ -200,5 +200,48 @@ def make_nonblocking_calls(expected, expect_blocking=False, num_returns=1):
     make_nonblocking_calls({"2": 2}, num_returns=2)
 
 
+def test_controller_recover_initializing_actor(serve_instance):
```
[the remark this answers is not in the document]
Comment (sihanwang41) [reply python/ray/serve/tests/test_controller_recovery.py:203]: yes!

File: python/ray/serve/tests/test_controller_recovery.py (hunk 1 above)
Comment (sihanwang41) [reply python/ray/serve/tests/test_controller_recovery.py:219]: I think it make no difference for this test case.

File: python/ray/serve/tests/test_controller_recovery.py [hunk 6]
```diff
@@ -200,5 +200,48 @@ def make_nonblocking_calls(expected, expect_blocking=False, num_returns=1):
...
+            if name in actor["name"]:
+                return actor["name"]
+        return None
+
+    actor_tag = get_actor_tag(V1.name)
+
+    ray.kill(serve.context._global_client._controller, no_restart=False)
+
+    # Let the actor proceed initialization
+    signal.send.remote()
```
[the remark this answers is not in the document]
Comment (sihanwang41) [reply python/ray/serve/tests/test_controller_recovery.py:239]: No need, if controller is not able to recover, it will fail with `client._wait_for_deployment_healthy`. The test is still valid for testing this scenario.

File: python/ray/serve/_private/replica.py [hunk 7]
```diff
@@ -129,6 +129,9 @@ async def __init__(
                 controller_name, namespace=SERVE_NAMESPACE
             )
 
+            # Indicate the replica finish the initialization.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/replica.py:132]:
```suggestion
            # Indicates whether the replica has finished initializing.
```

File: python/ray/serve/tests/test_controller_recovery.py [hunk 8]
```diff
@@ -200,5 +200,47 @@ def make_nonblocking_calls(expected, expect_blocking=False, num_returns=1):
...
+
+    @ray.remote
+    def pending_init_indicator():
+        ray.get(signal2.wait.remote())
+        return True
+
+    @serve.deployment
+    class V1:
+        def __init__(self):
+            signal2.send.remote()
```
Comment (shrekris-anyscale) [inline python/ray/serve/tests/test_controller_recovery.py:218]:
```suggestion
            ray.get(signal2.send.remote())
```

File: python/ray/serve/tests/test_controller_recovery.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/tests/test_controller_recovery.py:238]: Oh, I meant if the replica is able to fully initialize before the controller runs its own recovery process, then the code won't run into issue of `self.replica` being `None`, right? In that case we aren't testing the bug.

File: python/ray/serve/_private/replica.py [hunk 9]
```diff
@@ -243,9 +247,13 @@ async def reconfigure(
...
                 await self.replica.reconfigure(user_config)
 
-            return self.get_metadata()
+            return await self.get_metadata()
 
-        def get_metadata(self) -> Tuple[DeploymentConfig, DeploymentVersion]:
+        async def get_metadata(
+            self,
+        ) -> Tuple[DeploymentConfig, DeploymentVersion]:
+            # Wait for replica initialization finish
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/replica.py:255]:
```suggestion
            # Wait for replica initialization to finish
```

File: python/ray/serve/tests/test_controller_recovery.py (hunk 1 above)
Comment (zcin) [reply python/ray/serve/tests/test_controller_recovery.py:219]: Is the controller able to call the replica during recovery while the replica is stuck on `ray.get()`, since its a blocking call?

**Review by shrekris-anyscale — COMMENTED** [review]
Nice work! Could you give some more detail on why this error appears when the controller is dead and a replica is initializing? I don't understand why that causes `self.replica` to be `None`.

File: python/ray/serve/_private/deployment_state.py [hunk 10]
```diff
@@ -497,7 +497,10 @@ def check_ready(self) -> Tuple[ReplicaStartupStatus, Optional[DeploymentVersion]
                     self._allocated_obj_ref
                 )
             except Exception:
-                logger.exception(f"Exception in deployment '{self._deployment_name}'")
+                logger.exception(
+                    f"Exception in replica '{self._replica_tag}', "
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:501]: it'll get logged already because we're using `logger.exception`

File: python/ray/serve/_private/replica.py [hunk 11]
```diff
@@ -129,6 +129,9 @@ async def __init__(
                 controller_name, namespace=SERVE_NAMESPACE
             )
 
+            # Indicates whether the replica has finished initializing.
+            self.init_finish_event = asyncio.Event()
```
Comment (edoakes) [inline python/ray/serve/_private/replica.py:133]: nit: use `_` prefix for private variables
```suggestion
            self._init_finish_event = asyncio.Event()
```

**Review by edoakes — APPROVED** [review]

**Review by shrekris-anyscale — APPROVED** [review]
Great work on the changes!

**Review by zcin — APPROVED** [review]

**Comment by edoakes** [comment]
@sihanwang41 a serve AIR integration test is failing, maybe related?

**Comment by sihanwang41** [comment]
@edoakes it is not related, Cindy's pr https://github.com/ray-project/ray/pull/34039 also fails with same reason. looks like it is a separate issue.

**Comment by sihanwang41** [comment]
^^ @zhe-thoughts observe this bug from the release test, here is the fix.

**Comment by sihanwang41** [comment]
master is unfrozen, no need to have Zhe approval right now. ready to merge
